core/modules/replica/IndexJob.h

```diff
@@ -0,0 +1,292 @@
+    Destination        destination()     const { return _destination; }
+    std::string const& destinationPath() const { return _destinationPath; }
+
+    /**
+     * Return the combined result of the operation
+     *
+     * @note:
+     *  The method should be invoked only after the job has finished (primary
+     *  status is set to Job::Status::FINISHED). Otherwise exception
+     *  std::logic_error will be thrown
+     *
+     * @return
+     *   the data structure to be filled upon the completion of the job.
+     *
+     * @throws std::logic_error
+     *   if the job didn't finished at a time when the method was called
+     */
+    IndexJobResult const& getResultData() const;
+
+    std::list<std::pair<std::string,std::string>> extendedPersistentState() const final;
+    std::list<std::pair<std::string,std::string>> persistentLogData() const final;
+
+protected:
+
+    void startImpl(util::Lock const& lock) final;
+
+    void cancelImpl(util::Lock const& lock) final;
+
+    void notify(util::Lock const& lock) final;
+
+private:
+
+    IndexJob(std::string const& database,
+             bool hasTransactions,
+             uint32_t transactionId,
+             bool allWorkers,
+             Destination destination,
+             std::string const& destinationPath,
+             Controller::Ptr const& controller,
+             std::string const& parentJobId,
+             CallbackType const& onFinish,
+             Job::Options const& options);
+
+    /**
+     * The callback function to be invoked on a completion of requests
+     * targeting workers.
+     */
+    void _onRequestFinish(IndexRequest::Ptr const& request);
+
+    /**
+     * Extract data from the successfully completed requests. The completion
+     * state of the request will be evaluated by the method.
+     *
+     * @param lock
+     *   on the mutex Job::_mtx to be acquired for protecting the object's state
+     *
+     * @param request
+     *   the request to extract the data to be processed
+     */
+    void _processRequestData(util::Lock const& lock,
+                             IndexRequest::Ptr const& request);
+
+    /**
+     * Lunch a batch of requests with a total number not to exceed the specified
+     * limit.
+     *
+     * @param lock
+     *   on the mutex Job::_mtx to be acquired for protecting the object's state
+     *
+     * @param worker
+     *   the name of a worker the requests to be sent to
+     * 
+     * @param maxRequests
+     *   the maximum number of requests to be launched
+     *
+     * @return a collection of requests launched
+     */
+    std::list<IndexRequest::Ptr> _launchRequests(util::Lock const& lock,
+                                                 std::string const& worker,
+                                                 size_t maxRequests=1);
+
+    /**
+     * Roll back a database transaction should the one be still open
+     * for destination TABLE. The method won't have any effect for other
```

Comment:
I'm not sure what this means "Roll back a database transaction should the one be still open for destination TABLE."

Reply:
This is a clean up method which ensures no partially-ingest data is left in the corresponding MySQL partition in case of any failure while requesting/receiving or ingesting rows into the secondary index tables. The method is called from contexts where problems could be identifier, and, as the last resort, from the destructor of the class.
And `TABLE` is mentioned because this class supports 4 destinations for the secondary index data harvested from workers: `DISCARD` - request/receive and simply throw away the data, `FILE` - append all data (`TSV`/`CSV`) into a single file, `FOLDER` - put each chunk's data into a dedicated file stored in the specified folder, and `TABLE` - for actual MySQL table (usually in `qservMeta`). These destinations are explained at:
```
class IndexJob : public Job  {
public:
    ...
    /// Possible destinations where the harvested data would go
    enum Destination {
        DISCARD,    // do nothing with the data
        FILE,       // write all data into a file
        FOLDER,     // write each chunk's data as a separate file at a folder
        TABLE       // write into the specified or standard "secondary index" table
    };
```